device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/InternalClient.java

```diff
@@ -184,6 +184,33 @@ public void sendEventAsync(Message message, IotHubEventCallback callback, Object
         deviceIO.sendEventAsync(message, callback, callbackContext, this.config.getDeviceId());
     }
 
+    /**
+     * ASynchronously sends a batch of messages to the IoT Hub
+     * HTTPS messages will be sent in a single batch and MQTT and AMQP messages will be sent individually.
+     * Maximum payload size for HTTPS is 255KB
+     *
+     * @param messages the list of message to be sent.
+     * @param callback the callback to be invoked when a response is received.
+     * Can be {@code null}.
+     * @param callbackContext a context to be passed to the callback. Can be
+     * {@code null} if no callback is provided.
+     *
+     * @throws IllegalArgumentException if the message provided is {@code null}.
+     * @throws IllegalStateException if the client has not been opened yet or is
+     * already closed.
+     */
+    public void sendEventBatchAsync(Set<Message> messages, IotHubEventCallback callback, Object callbackContext)
```

Comment:
The ordering of the messages needs to be preserved, so maybe we should use a `List` instead? Or add a comment that the customer should supply an ordered implementation of `Set`?

Reply:
Does the ordering need to be preserved? I thought the whole point was that they were all delivered simultaneously?

Reply:
yes, they would be delivered simultaneously, but still as a batch containing the individual messages in some order -> if I add them as 1->2->3 into a batch; then on service send, once I receive a batch and iterate over it, the order should still be 1->2->3.

Reply:
I will make a note of that in the comments.